src/metadataExt.ts

```diff
@@ -0,0 +1,261 @@
+                                const { visualizationWidget } = widgets[index];
+
+                                return visualizationWidget.meta.uri;
+                            }
+                        }
+                    })
+                    .filter(links => links !== undefined)
+                    .join(",");
+            }),
+        );
+
+        const clonedLayout = this.updateLayout(dashboard, widgetLinks, widgets);
+
+        const updatedDashboardLink: string = await this.updateFilterContext(
+            projectID,
+            dashboardID,
+            dashboard,
+            objectsUsingDashboard,
+        );
+        const { analyticalDashboard } = await this.metadataModule.getObjectDetails(updatedDashboardLink);
+
+        const { uri }: { uri: string } = await this.metadataModule.updateObject(projectID, dashboardID, {
+            analyticalDashboard: {
+                ...analyticalDashboard,
+                content: {
+                    ...analyticalDashboard.content,
+                    layout: { ...analyticalDashboard.content.layout, fluidLayout: clonedLayout },
+                    widgets: await replaceWidgetsRef,
+                },
+            },
+        });
+
+        return uri;
+    }
+
+    private updateLayout(dashboard: any, widgetLinks: string[], widgets: any[]) {
+        return cloneDeepWith({ ...dashboard.content.layout.fluidLayout }, key => {
+            return widgetLinks
+                .map((link, index) => {
+                    const category = Object.keys(widgets[index])[0];
+                    if (key === link) {
+                        if (category === "kpi") {
+                            const { kpi } = widgets[index];
+                            return kpi.meta.uri;
+                        } else {
+                            const { visualizationWidget } = widgets[index];
+                            return visualizationWidget.meta.uri;
+                        }
+                    }
+                })
+                .filter(link => link !== undefined)[0];
+        });
+    }
+
+    private async updateFilterContext(
+        projectId: string,
+        dashboardID: string,
+        analyticalDashboard: any,
+        objectsUsingDashboard: any[],
+    ): Promise<string> {
+        const { link }: { link: string } = objectsUsingDashboard.filter((obj: any) =>
+            obj.category.match("filterContext"),
+        )[0];
+
+        const { filterContext } = await this.metadataModule.createObject(
+            projectId,
+            await this.metadataModule.getObjectDetails(link),
+        );
+
+        const { uri }: { uri: string } = await this.metadataModule.updateObject(projectId, dashboardID, {
+            analyticalDashboard: {
+                ...analyticalDashboard,
+                content: {
+                    ...analyticalDashboard.content,
+                    filterContext: filterContext.meta.uri,
+                },
+            },
+        });
```

Comment:
same as before, this should go away if you create new instance of dashboard at the very end of saveAs processing. until then you can pass around and work with dashboard object in memory